bus/contractlocking.go

```diff
@@ -0,0 +1,205 @@
+package bus
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"go.sia.tech/siad/types"
+	"lukechampine.com/frand"
+)
+
+// ErrAcquireContractTimeout is returned when the context passed in to
+// contractLocks.Acquire is closed before the lock can be acquired.
+var ErrAcquireContractTimeout = errors.New("acquiring the lock timed out")
+
+type contractLocks struct {
+	mu    sync.Mutex
+	locks map[types.FileContractID]*contractLock
+}
+
+type contractLock struct {
+	mu          sync.Mutex    // locks contractLock fields
+	lockChan    chan struct{} // locks the contract
+	lockedUntil time.Time
+	heldBy      uint64
```

Comment:
`heldByID` is slightly better I think, I originally thought it was a ref. counter or something